GameProfile field not found in TinyProtocol

- [X] This issue is not solved in a development build

**Describe the bug**
When using the [TinyProtocol class](https://github.com/dmulloy2/ProtocolLib/blob/master/TinyProtocol/src/main/java/com/comphenix/tinyprotocol/TinyProtocol.java#L67), at line 67 the `Cannot find field with type class com.mojang.authlib.GameProfile` has occurred.

**To Reproduce**
Steps to reproduce the behavior:
1. Add TinyProtocol and Reflection classes into your plugin project
2. Add a simple implementation like the following one:
```
package it.test.protocol;

import io.netty.channel.Channel;
import it.test.protocol.TinyProtocol;
import org.bukkit.entity.Player;
import org.bukkit.plugin.Plugin;

public class TestTinyProtocol extends TinyProtocol {

    public TestTinyProtocol(Plugin plugin) {
        super(plugin);
    }

    @Override
    public Object onPacketOutAsync(Player receiver, Channel channel, Object packet) {
        plugin.getLogger().info("OUT: " + packet.toString());
        return super.onPacketOutAsync(receiver, channel, packet);
    }

    @Override
    public Object onPacketInAsync(Player sender, Channel channel, Object packet) {
        plugin.getLogger().info("IN: " + packet.toString());
        return super.onPacketInAsync(sender, channel, packet);
    }
}
```

4. Add it to the plugin onLoad/onEnable (same behavior)
`this.tinyProtocol = new TestTinyProtocol(plugin);`
6. Start the server and wait until the error occurred.

**Exception**
```
Caused by: java.lang.IllegalArgumentException: Cannot find field with type class com.mojang.authlib.GameProfile
        at it.test.protocol.Reflection.getField(Reflection.java:179) ~[ParadiseLand.jar:?]
        at it.test.protocol.Reflection.getField(Reflection.java:177) ~[ParadiseLand.jar:?]
        at it.test.protocol.Reflection.getField(Reflection.java:177) ~[ParadiseLand.jar:?]
        at it.test.protocol.Reflection.getField(Reflection.java:123) ~[ParadiseLand.jar:?]
        at it.test.protocol.TinyProtocol.<clinit>(TinyProtocol.java:67) ~[ParadiseLand.jar:?]
        ... 14 more
```

This is because the `net.minecraft.network.protocol.login.PacketLoginInStart` from Spigot 1.19 no longer has a GameProfile field:
![image](https://github.com/dmulloy2/ProtocolLib/assets/9408687/806c7387-5082-41c1-adf2-43c1e24dc460)


**Expected behavior**
Correct TinyProtocol loading.

**Version Info**
**ProtocolLib**: not used in this case
**TinyProtocol**: the latest available class from the GitHub master branch.

**Additional context**
TinyProtocol works from 1.8.8 to 1.18.2 and does not work from: **1.19**

Thanks! 🥇

--- a/TinyProtocol/src/main/java/com/comphenix/tinyprotocol/TinyProtocol.java
+++ b/TinyProtocol/src/main/java/com/comphenix/tinyprotocol/TinyProtocol.java
@@ -64,7 +64,7 @@ public abstract class TinyProtocol {
 
     // Packets we have to intercept
     private static final Class<?> PACKET_LOGIN_IN_START = Reflection.getClass("{nms}.PacketLoginInStart", "net.minecraft.network.protocol.login.PacketLoginInStart");
-    private static final FieldAccessor<GameProfile> getGameProfile = Reflection.getField(PACKET_LOGIN_IN_START, GameProfile.class, 0);
+    private static final FieldAccessor<String> getPlayerName = new PlayerNameAccessor();
 
     // Speedup channel lookup
     private Map<String, Channel> channelLookup = new MapMaker().weakValues().makeMap();
@@ -493,6 +493,59 @@ public abstract class TinyProtocol {
     }
 
     /**
+     * Get the player name from the login start packet.
+     * This fixes the issue from 1.19 where the GameProfile field has been removed from this login packet.
+     *
+     * @author gamerover98
+     */
+    private static class PlayerNameAccessor implements FieldAccessor<String> {
+
+        private FieldAccessor<String> getPlayerName;
+        private FieldAccessor<GameProfile> getGameProfile;
+
+        PlayerNameAccessor() {
+            try {
+                this.getGameProfile = Reflection.getField(PACKET_LOGIN_IN_START, GameProfile.class, 0);
+            } catch (IllegalArgumentException illegalArgumentException) {
+                // nothing to do.
+            }
+
+            try {
+                //HOT-FIX for 1.19+
+                this.getPlayerName = Reflection.getField(PACKET_LOGIN_IN_START, String.class, 0);
+            } catch (IllegalArgumentException illegalArgumentException) {
+                // nothing to do.
+            }
+
+            if (getGameProfile == null && getPlayerName == null) {
+                throw new UnsupportedOperationException("The current server version is not supported by TinyProtocol");
+            }
+        }
+
+        @Override
+        public String get(Object target) {
+            if (getPlayerName != null) {
+                String playerName = getPlayerName.get(target);
+                return playerName.substring(0, Math.min(16, playerName.length()));
+            }
+
+            return getGameProfile.get(target).getName();
+        }
+
+        @Override
+        public void set(Object target, Object value) {
+            throw new UnsupportedOperationException("Not supported");
+        }
+
+        @Override
+        public boolean hasField(Object target) {
+            return getPlayerName != null
+                    ? getPlayerName.hasField(target)
+                    : getGameProfile.hasField(target);
+        }
+    }
+
+    /**
      * Channel handler that is inserted into the player's channel pipeline, allowing us to intercept sent and received packets.
      * 
      * @author Kristian
@@ -533,8 +586,7 @@ public abstract class TinyProtocol {
 
         private void handleLoginStart(Channel channel, Object packet) {
             if (PACKET_LOGIN_IN_START.isInstance(packet)) {
-                GameProfile profile = getGameProfile.get(packet);
-                channelLookup.put(profile.getName(), channel);
+                channelLookup.put(getPlayerName.get(packet), channel);
             }
         }
     }

